contrib/submit-job-v2/src/App/ProtocolForm.tsx

```diff
@@ -571,46 +661,280 @@ export default class ProtocolForm extends React.Component<IProtocolProps, IProto
+      case "HDFS":
+        preCommand.push([
+          "apt-get install -y git fuse golang",
+          "git clone --recursive https://github.com/Microsoft/hdfs-mount.git",
+          "cd hdfs-mount",
+          "make",
+          "cp hdfs-mount /bin",
+          "cd ..",
+          "rm -rf hdfs-mount",
+        ].join("&&"));
+        if (storageConfig.hostIP === undefined || storageConfig.remotePath === undefined) {
+          preCommand = [];
+          break;
+        }
+        preCommand.push(`(hdfs-mount ${storageConfig.hostIP}:${storageConfig.port} ${mountPath} &)`);
+        preCommand.push(`sleep 5`);
+        mountPath += storageConfig.remotePath;
+        break;
+      case "NFS":
+        if (storageConfig.hostIP === undefined || storageConfig.remotePath === undefined) {
+          preCommand = [];
+          break;
+        }
+        preCommand.push("apt-get install -y nfs-common");
+        preCommand.push(`mount -t nfs4 ${storageConfig.hostIP}:${storageConfig.remotePath} ${mountPath}`);
+        break;
+      default:
+        preCommand = [];
+        break;
+    }
+    if (preCommand.length === 0) {
+      alert("Please complete the storage config!");
+      return [];
+    }
+    preCommand.push(`export TB_ROOT=${mountPath}`);
+    preCommand.push("`#TENSORBOARD_STORAGE_END`");
+    return preCommand;
+  }
+
+  private injectCommand = async () => {
+    const preCommand = await this.generatePreCommand();
+    if (preCommand.length === 0) {
+      return false;
+    }
     const protocol = yaml.safeLoad(this.state.protocolYAML);
-    const randomStr = Math.random().toString(36).slice(-8);
-    const tensorBoardName = `TensorBoard_${randomStr}`;
-    const tensorBoardImage = `tensorBoardImage_${randomStr}`;
-    const tensorBoardPort = `tensorBoardPort_${randomStr}`;
-    let length = 0;
-    if (protocol.hasOwnProperty("prerequisites")) {
-      length = protocol.prerequisites.length;
+    if (protocol.hasOwnProperty("taskRoles")) {
+      const obj = protocol.taskRoles;
+      Object.keys(obj).forEach((key) => {
+        obj[key].commands = preCommand.concat(obj[key].commands);
+      });
+    }
+    this.setState({
+      protocol,
+      protocolYAML: yaml.safeDump(protocol),
+    });
+  }
+
+  private addTensorBoardConfig = async () => {
+    const protocol = this.state.protocol;
+    if (protocol === undefined) {
+      return;
+    }
+    const tensorBoardConfig = this.state.tensorBoardConfig;
+    tensorBoardConfig.randomStr = Math.random().toString(36).slice(2, 10);
+    if (Object.prototype.hasOwnProperty.call(protocol, "extras")) {
+      protocol.extras.tensorBoard = tensorBoardConfig;
     } else {
+      protocol.extras = { tensorBoard: tensorBoardConfig };
+    }
+    this.setState({
+      protocol,
+      protocolYAML: yaml.safeDump(protocol),
+      tensorBoardConfig,
+    });
+    await this.addTensorBoardTaskRole();
+    await this.injectCommand();
+  }
+
+  private addTensorBoardTaskRole = async () => {
```

Comment:
does `addTensorBoardTaskRole` need to be `async`? doesn't wait `Promise` in it.

Reply:
OK, I will rebuild the code.